src/ripple/app/tx/impl/AMMDeposit.cpp

```diff
@@ -0,0 +1,528 @@
+        {
+            if (asset2In)
+                return equalDepositLimit(
+                    sb,
+                    ammAccountID,
+                    asset1,
+                    asset2,
+                    lptAMMBalance,
+                    *asset1In,
+                    *asset2In);
+            else if (lpTokensDeposit)
+                return singleDepositTokens(
+                    sb,
+                    ammAccountID,
+                    asset1,
+                    lptAMMBalance,
+                    *lpTokensDeposit,
+                    tfee);
+            else if (ePrice)
+                return singleDepositEPrice(
+                    sb,
+                    ammAccountID,
+                    asset1,
+                    *asset1In,
+                    lptAMMBalance,
+                    *ePrice,
+                    tfee);
+            else
+                return singleDeposit(
+                    sb, ammAccountID, asset1, lptAMMBalance, *asset1In, tfee);
+        }
+        else if (lpTokensDeposit)
+            return equalDepositTokens(
+                sb,
+                ammAccountID,
+                asset1,
+                asset2,
+                lptAMMBalance,
+                *lpTokensDeposit);
+        // should not happen.
+        JLOG(j_.error()) << "AMM Deposit: invalid options.";
+        return std::make_pair(tecAMM_FAILED_DEPOSIT, STAmount{});
+    }();
+
+    if (result == tesSUCCESS && depositedTokens != beast::zero)
+    {
+        ammSle->setFieldAmount(
+            sfLPTokenBalance, lptAMMBalance + depositedTokens);
+        sb.update(ammSle);
+    }
+
+    return {result, result == tesSUCCESS};
+}
+
+TER
+AMMDeposit::doApply()
+{
+    // This is the ledger view that we work against. Transactions are applied
+    // as we go on processing transactions.
+    Sandbox sb(&ctx_.view());
+
+    // This is a ledger with just the fees paid and any unfunded or expired
+    // offers we encounter removed. It's used when handling Fill-or-Kill offers,
+    // if the order isn't going to be placed, to avoid wasting the work we did.
+    Sandbox sbCancel(&ctx_.view());
+
+    auto const result = applyGuts(sb);
+    if (result.second)
+        sb.apply(ctx_.rawView());
+    else
+        sbCancel.apply(ctx_.rawView());
+
+    return result.first;
+}
+
+std::pair<TER, STAmount>
+AMMDeposit::deposit(
+    Sandbox& view,
+    AccountID const& ammAccount,
+    STAmount const& asset1Deposit,
+    std::optional<STAmount> const& asset2Deposit,
+    STAmount const& lpTokensDeposit)
+{
+    // Check account has sufficient funds
+    auto balance = [&](auto const& asset) {
+        return accountHolds(
+                   view,
+                   account_,
+                   asset.issue().currency,
+                   asset.issue().account,
+                   FreezeHandling::fhZERO_IF_FROZEN,
+                   ctx_.journal) >= asset;
+    };
+
+    // Deposit asset1Deposit
+    if (!balance(asset1Deposit))
```

Comment:
I don't quite understand this. Why don't we check if it's greater then the amount to deposit? Why just check for zero?